### DISCUSSION TITLE
How to specify CSV column names for charts

### DISCUSSION THREAD

**Discussion by amruthkiran94** [TerriaJS/terriajs#6380] [Q&A]
Hey!
So I've been tinkering around with v8 a bit where I'm trying to display a CSV file as charts based on each coordinate clicked. I have the coordinates displayed but I'm not able to figure out how to specify the column names from the CSV that has to be charted. Sharing the code below:
`{
    "catalog": [
      {
            "name": "NAES",
            "type": "csv",
            "url": "build/TerriaJS/test/csv/naesall.csv",
            "featureInfoTemplate" : {
              "name": "{{site}} | Viral Load:{{viral_load}}",
              "template": "<chart src='build/TerriaJS/test/csv/naesall.csv' column-names='Time,Viral Load' column-units='Date,VL'></chart>"
            }
      }
    ]
  }`

I've tried using `x-column={{date}} x-column={{viral_load}}` and `x-column='date' x-column= 'viral_load' `. In both instances I get a `No chartable data found` error. Any help would be highly appreciated.

Thanks!

**Comment by nf-s** [comment]
Hi @amruthkiran94 

Is this CSV time-series? If so the chart should automatically be shown for the selected feature.

For example http://ci.terria.io/main/#share=s-iC6I6sm2MzhLNsJYm4X0vbrjnWk

Which is using http://ci.terria.io/main/build/TerriaJS/test/csv/lat_lon_enum_date_id.csv

--------

There is info about `featureInfoTemplate` and charts here https://github.com/TerriaJS/terriajs/blob/main/doc/connecting-to-data/customizing-data-appearance/feature-info-template.md#time-series-charts

As for setting `x-column` to a specific column name - you should use `x-column='some_col_name'`

There is also info about chart attributes here

https://github.com/TerriaJS/terriajs/blob/fc33525049a701c3d1876b9ff9fbffc1c65e4310/lib/ReactViews/Custom/ChartCustomComponent.ts#L29-L92

Attribute names are in kebab-case - eg `previewXLabel` becomes `preview-x-label`

If you are able to provide a subset of your dataset I can have a closer look

Cheers,
Nick

**Comment by amruthkiran94** [comment]
Thank you so much for the response @nf-s. I was also exploring the National Map's "Wind" dataset linked [here](https://nationalmap.gov.au/#share=s-au098Z66G0TN89TEQPfXOKKTTj7). Basically I'm trying to build a real-time and dynamic charting behavior using a PostgreSQL API (PostgREST) that is linked to a catalogue file. I've already had a look at a similar question posted [here ](https://github.com/TerriaJS/terriajs/discussions/5028) as the basis of my work.

The file, based on point selection: 
 ```
   {
      "id":"1b",
      "name":"CSV as API (Local)",
      "url":"data/latlong_data.csv",
      "type":"csv",
      "cacheDuration":"5m",
      "featureInfoTemplate":{
         "template":"<chart title='{{name}}' sources='http://localhost:3000/sample?name=eq.{{name}}' column-names='Time,Value' column-units='Date,Value'></chart>"
      }
}
``` 
The sources parameter requests the table to chart but its response type is JSON 
```
[{"duid":"2","name":"Point_B","value":200,"lat":12.99694736,"lon":77.59832382}]
```
TerriaMap doesn't seem to be accepting this response type. Is there any way to accept a CSV via the `sources` parameter itself or do I have to do something else?

**Reply by nf-s** [reply]
Currently, there is no way to set http `Accept` headers for `sources`.

We do have a new `ApiTableCatalogItem` which can transform JSON into tabular data.
This can be used by replacing `chart` with `api-chart`. 

`api-chart` has a `api-table-catalog-item-json` attribute which can be used to set trait values - see https://docs.terria.io/guide/connecting-to-data/catalog-type-details/api-table/ for trait docs

I don't have an example of `api-chart` ready - but it would be something like

```json
"template":"<api-chart title='{{name}}' api-table-catalog-item-json='{\"apis\":[{\"url\":\"http://localhost:3000/sample?name=eq.{{name}}\"}]}' column-names='Time,Value' column-units='Date,Value'></api-chart>"
```

where `api-table-catalog-item-json` is the following JSON blob stringified

```json
{
  "apis": [
    {
      "url": "http://localhost:3000/sample?name=eq.{{name}}"
    }
  ]
}
```

All of this is quite new and experimental, so it might be unpredictable

**Reply by amruthkiran94** [reply]
Thanks @nf-s I've been trying out your suggestion, here's what I found
```
{
  "id":"1c",
  "name":"CSV as API-Chart (Local)",
  "url":"data/latlong_data.csv",
  "type":"api-table",
  "cacheDuration":"5m",
  "featureInfoTemplate":{
    "template":"<api-chart title='{{name}}' api-table-catalog-item-json='{\"apis\":[{\"url\":\"http://localhost:3000/sample\"}]}' column-names='Time,Value' column-units='Date,Value'></api-chart>"  
  }
}
```
Based on what I could understand, I've set `type` as `api-table` and placed the `template` as you suggested but it looks like `featureInfoTemplate` isn't a trait for `api-table`. How would I create a catalouge item that can also support the `featureInfoTemplate` for `api-table`?

Edit:
I've also found this https://github.com/TerriaJS/terriajs/pull/5373#issuecomment-809950067 where an example was provided by @KeyboardSounds 
```
{
      "name": "Pedestrian Traffic",
      "type": "api-table",
      "refreshInterval": 60,
      "removeDuplicateRows": true,
      "idKey": "sensor_id",
      "dateTimeKey": "date_time",
      "defaultStyle": {
        "time": {
          "idColumns": ["sensor_id"]
        }
      },
      "apis": [
        {
          "apiUrl": "https://data.melbourne.vic.gov.au/resource/d6mv-s43h.json",
          "keyToColumnMapping": [
            { 
              "columnName": "Total in both directions",
              "keyInApiResponse": "total_of_directions"
            },
            { 
              "columnName": "Time",
              "keyInApiResponse": "date_time"
            },
            { 
              "columnName": "id",
              "keyInApiResponse": "sensor_id"
            }
          ]
        },
        {
          "apiUrl": "https://data.melbourne.vic.gov.au/resource/h57g-5234.json",
          "kind": "PER_ID",
          "keyToColumnMapping": [
            { 
              "columnName": "latitude",
              "keyInApiResponse": "latitude"
            },
            { 
              "columnName": "longitude",
              "keyInApiResponse": "longitude"
            }
          ]
        }
      ]
    }
```
I'm not sure how to implement this for my use-case, as some of the attributes like `keyToColumnMapping` has been changed (atleast as per the latest V8 docs) . Any help would be amazing!

Thanks!

**Reply by nf-s** [reply]
Hi @amruthkiran94 

Sorry I didn't realise that `ApiTableCatalogItem` didn't have `featureInfoTemplate` yet!

We have two PRs to add that functionality - so hopefully we will get them merged soon
- https://github.com/TerriaJS/terriajs/pull/5660
- https://github.com/TerriaJS/terriajs/pull/6415

Instead of using `keyToColumnMapping` - you can set column titles using `TableColumnTraits`.

This is an updated version of the `"Pedestrian Traffic"` catalog item

```json
{
  "type": "api-table",
  "name": "City of Melbourne Pedestrian Traffic Counter",
  "description": "The City of Melbourne Pedestrian Counting System data has been obtained from the City of Melbourne Open data platform: www.data.melbourne.vic.gov.au to demonstrate the directional movement of pedestrians across the city.\n\nThe counters generate a per minute reading on each of the pedestrian traffic sensors which are then displayed as near-live data.\n\nTo find out more about the City of Melbourne's Pedestrian Counting System or to seek access to their data, please click:\n\nhttp://www.pedestrian.melbourne.vic.gov.au/#date=27-09-2021&time=1",
  "apis": [
    {
      "url": "https://data.melbourne.vic.gov.au/resource/d6mv-s43h.json?$limit=50000"
    },
    {
      "kind": "PER_ID",
      "url": "https://data.melbourne.vic.gov.au/resource/h57g-5234.json"
    }
  ],
  "idKey": "sensor_id",
  "columns": [
    {
      "name": "total_of_directions",
      "title": "Total in both directions"
    },
    {
      "name": "date_time",
      "title": "Time"
    },
    {
      "name": "latitude",
      "title": "latitude"
    },
    {
      "name": "longitude",
      "title": "longitude"
    },
    {
      "name": "sensor_id",
      "title": "Sensor ID"
    },
    {
      "name": "installation_date",
      "title": "Installation Date"
    },
    {
      "name": "sensor_description",
      "title": "Sensor Description",
      "type": "hidden"
    }
  ],
  "activeStyle": "total_of_directions",
  "defaultStyle": {
    "time": {
      "spreadStartTime": true,
      "timeColumn": "date_time"
    }
  },
  "refreshInterval": 60,
  "removeDuplicateRows": true
}
```

**Reply by amruthkiran94** [reply]
Thank you so much for this @nf-s! Is there a complete catalogue item I could possibly access? Would love to see the `chart` and `featureInfoTemplate` details as well.
